applications/sudokumb/GameLib/GameBoard.cs

```diff
@@ -0,0 +1,293 @@
+                    _board.Substring(i + 3, 3), _board.Substring(i + 6, 3));
+                if (i == 18 || i == 45)
+                    s.AppendLine("---+---+---");
+            }
+            return s.ToString();
+        }
+
+        /// <summary>
+        /// Parse a GameBoard from user input or the like.  The format is
+        /// different from the pretty output, but very easier to enter by hand.
+        /// Digits for digits, and anything else for blank spaces.
+        /// Whitespace is ignored.  Example:
+        /// 1 2 3   5 . .   7 8 9
+        /// . . .   . . .   . . .
+        /// . . .   . . .   . . .
+        ///
+        /// . . .   4 . .   . . .                
+        /// . 7 .   . 5 .   . . .
+        /// . . .   . . 6   2 . .
+        ///
+        /// . . 1   . . .   . . .
+        /// . 5 .   3 . .   . . .
+        /// 3 . .   . . .   1 . .
+        /// </summary>
+        /// <param name="input">An input stream.  Try Console.OpenStandordInput()</param>
+        public static GameBoard ParseHandInput(Stream input)
+        {
+            var reader = new StreamReader(input);
+            // Parse the first line to see if it's a simple format or pretty
+            // format.
+            StringBuilder board = new StringBuilder();
+            do
+            {
+                int n = reader.Read();
+                if (n < 0)
+                {
+                    throw new ArgumentException("Input stream did contain a full board.");
+                }
+                char c = (char)n;
+                if ("123456789".Contains(c))
+                {
+                    board.Append(c);
+                }
+                else if (Char.IsWhiteSpace(c))
+                {
+                    continue;
+                }
+                else
+                {
+                    board.Append(" ");
+                }
+            } while (board.Length < 81);
+            return GameBoard.Create(board.ToString());
+        }
+
+        public static GameBoard ParseHandInput(string inputString)
+        {
+            var stream = new MemoryStream(Encoding.UTF8.GetBytes(inputString));
+            return ParseHandInput(stream);
+        }
+
+        public string ToHandInputString()
+        {
+            StringBuilder s = new StringBuilder();
+            string b = _board.Replace(' ', '.');
+            int i = 0;
+            // Line
+            while (i < b.Length)
+            {
+                // Line
+                s.AppendFormat("{0} {1} {2}   {3} {4} {5}   {6} {7} {8}\n",
+                    b[i++], b[i++], b[i++],
+                    b[i++], b[i++], b[i++],
+                    b[i++], b[i++], b[i++]);
+                if (i == 27 || i == 54)
+                {
+                    s.AppendLine();
+                }
+            }
+            return s.ToString();
+        }
+    }
+
+    public class BadGameBoardException : Exception
```

Comment:
I read the article a couple times, and I don't see the benefit of implementing this exception the way thee article describes.  I see it would take many more lines of code that are never invoked, and they would need to be maintained.